Source/Core/Core/IOS/USB/Emulated/Infinity.cpp

```diff
@@ -0,0 +1,987 @@
+                                              m_player_two_ability_two;
+
+  figure.inf_file = std::move(in_file);
+  memcpy(figure.data.data(), buf, figure.data.size());
+  figure.present = true;
+  if (figure.order_added == 255)
+  {
+    figure.order_added = m_figure_order;
+    m_figure_order++;
+  }
+  order_added = figure.order_added;
+
+  while (position > 3)
+    position = position - 2;
+
+  std::array<u8, 32> figure_change_response = {0xab, 0x04, position, 0x09, order_added, 0x00};
+  figure_change_response[6] = GenerateChecksum(figure_change_response.data(), 6);
+  m_figure_added_removed_response.push(figure_change_response);
+
+  return FindFigure(number);
+}
+
+void InfinityBase::RemoveFigure(u8 position)
+{
+  std::lock_guard lock(m_infinity_mutex);
+  InfinityFigure& figure = position == 0x01 ? m_hexagon :
+                           position == 0x02 ? m_player_one :
+                           position == 0x03 ? m_player_two :
+                           position == 0x04 ? m_player_one_ability_one :
+                           position == 0x05 ? m_player_two_ability_one :
+                           position == 0x06 ? m_player_one_ability_two :
+                                              m_player_two_ability_two;
+  figure.present = false;
+
+  while (position > 3)
+    position = position - 2;
+
+  std::array<u8, 32> figure_change_response = {0xab, 0x04, position, 0x09, figure.order_added,
+                                               0x01};
+  figure_change_response[6] = GenerateChecksum(figure_change_response.data(), 6);
+  m_figure_added_removed_response.push(figure_change_response);
+}
+
+bool InfinityBase::CreateFigure(const std::string& file_path, u16 figure_num)
+{
+  File::IOFile inf_file(file_path, "w+b");
+  if (!inf_file)
+  {
+    return false;
+  }
+  InfinityFigureInfo figure = {InfinityFigureType::Character, 0};
+  for (auto it = list_infinity_figures.begin(); it != list_infinity_figures.end(); it++)
+  {
+    if (it->second.figure_number == figure_num)
+    {
+      figure = it->second;
+      break;
+    }
+  }
+  if (figure.figure_number == 0)
+  {
+    return false;
+  }
+  std::array<u8, INFINITY_NUM_BLOCKS * INFINITY_BLOCK_SIZE> file_data{};
+  u32 first_block = 0x17878E;
+  u32 other_blocks = 0x778788;
+  for (s8 i = 2; i >= 0; i--)
+  {
+    file_data[0x38 - i] = u8((first_block >> i * 8) & 0xFF);
+  }
+  for (u32 index = 1; index < 0x05; index++)
+  {
+    for (s8 i = 2; i >= 0; i--)
+    {
+      file_data[((index * 0x40) + 0x38) - i] = u8((other_blocks >> i * 8) & 0xFF);
+    }
+  }
+  std::vector<u8> sha1_calc = {SHA1_CONSTANT.begin(), SHA1_CONSTANT.end() - 1};
+  std::array<u8, 16> uid_data = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x89, 0x44, 0x00, 0xC2};
+  Common::Random::Generate(&uid_data[0], 7);
+  for (s8 i = 0; i < 7; i++)
+  {
+    sha1_calc.push_back(uid_data[i]);
+  }
+  std::array<u8, 16> figure_data = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
```

Comment:
Consider moving the figure data generation logic into a separate function. As it stands this function does way too many different things and has too many variables - making it rather difficult to keep track of everything it does.